src/corehost/cli/json_parser.cpp

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// These are only used by rapidjson/error/en.h to declare the error messages,
+// and have to be set to these values before any files are included.  They're
+// defined here because it's the only place that calls GetParseError().
+#undef RAPIDJSON_ERROR_CHARTYPE
+#undef RAPIDJSON_ERROR_STRING
+#define RAPIDJSON_ERROR_CHARTYPE pal::char_t
+#define RAPIDJSON_ERROR_STRING(x) _X(x)
+
+#include "json_parser.h"
+#include "rapidjson/error/en.h"
+#include "utils.h"
+#include <cassert>
+#include <cstdint>
+
+namespace {
+
+// Try to match 0xEF 0xBB 0xBF byte sequence (no endianness here.)
+std::streampos get_utf8_bom_length(pal::istream_t& stream)
+{
+    if (stream.eof())
+    {
+        return 0;
+    }
+
+    auto peeked = stream.peek();
+    if (peeked == EOF || ((peeked & 0xFF) != 0xEF))
+    {
+        return 0;
+    }
+
+    unsigned char bytes[3];
+    stream.read(reinterpret_cast<char*>(bytes), 3);
+    if ((stream.gcount() < 3) || (bytes[1] != 0xBB) || (bytes[2] != 0xBF))
+    {
+        return 0;
+    }
+
+    return 3;
+}
+
+void get_line_column_from_offset(const std::vector<char>& json, size_t offset, int *line, int *column)
+{
+    *line = *column = 1;
+
+    for (size_t i = 0; i < offset; i++)
```

Comment:
Nit: can you add an assert that offset fits into the vector? Just paranoia - and also acts as a documentation kind of thing.

Reply:
>and also acts as a documentation kind of thing.

I do this on a regular basis. Asserts are easy to change and I find that they are great for giving context to future developers about how I was thinking about a feature. The assert doesn't necessarily mean there is an issue, rather it can mean "I only verified the following and anything more should probably be thought about and/or tested". /cc @elinor-fung.